Language/Haskell/GhcMod/Utils.hs

```diff
@@ -156,3 +160,41 @@ canonFilePath f = do
   e <- doesFileExist p
   when (not e) $ error $ "canonFilePath: not a file: " ++ p
   return p
+
+withMappedFile :: (IOish m, GmState m, GmEnv m) =>
+                  forall a. FilePath -> (FilePath -> m a) -> m a
+withMappedFile file action = getCanonicalFileNameSafe file >>= lookupMMappedFile >>= runWithFile
+  where
+    runWithFile (Just (RedirectedMapping to)) = action to
+    runWithFile (Just (MemoryMapping (Just src))) = do
+      crdl <- cradle
+      (fp,hndl) <- liftIO $ openTempFile (cradleTempDir crdl) (takeFileName file)
+      liftIO $ hPutStr hndl src
+      liftIO $ hClose hndl
+      result <- action fp
+      liftIO $ removeFile fp
+      return result
+    runWithFile _ = action file
+
+getCanonicalFileNameSafe :: (IOish m, GmEnv m) => FilePath -> m FilePath
```

Comment:
So as far as I understand you need this because the redirected file might not actually exist yet, right? I don't know if that would be a good idea but how about just requiring it to already exist? (Like even just as an empty file)